homeassistant/helpers/condition.py

```diff
@@ -94,6 +103,95 @@
 )
 
 
+CONDITION_DESCRIPTION_CACHE: HassKey[dict[str, dict[str, Any] | None]] = HassKey(
+    "condition_description_cache"
+)
+CONDITION_PLATFORM_SUBSCRIPTIONS: HassKey[
+    list[Callable[[set[str]], Coroutine[Any, Any, None]]]
+] = HassKey("condition_platform_subscriptions")
+CONDITIONS: HassKey[dict[str, str]] = HassKey("conditions")
+
+
+# Basic schemas to sanity check the condition descriptions,
+# full validation is done by hassfest.conditions
+_FIELD_SCHEMA = vol.Schema(
+    {},
+    extra=vol.ALLOW_EXTRA,
+)
+
+_CONDITION_SCHEMA = vol.Schema(
+    {
+        vol.Optional("fields"): vol.Schema({str: _FIELD_SCHEMA}),
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+def starts_with_dot(key: str) -> str:
+    """Check if key starts with dot."""
+    if not key.startswith("."):
+        raise vol.Invalid("Key does not start with .")
+    return key
+
+
+_CONDITIONS_SCHEMA = vol.Schema(
+    {
+        vol.Remove(vol.All(str, starts_with_dot)): object,
+        cv.slug: vol.Any(None, _CONDITION_SCHEMA),
+    }
+)
+
+
+async def async_setup(hass: HomeAssistant) -> None:
+    """Set up the condition helper."""
+    hass.data[CONDITION_DESCRIPTION_CACHE] = {}
+    hass.data[CONDITION_PLATFORM_SUBSCRIPTIONS] = []
+    hass.data[CONDITIONS] = {}
+    await async_process_integration_platforms(
+        hass, "condition", _register_condition_platform, wait_for_platforms=True
+    )
+
+
+@callback
+def async_subscribe_platform_events(
+    hass: HomeAssistant,
+    on_event: Callable[[set[str]], Coroutine[Any, Any, None]],
+) -> Callable[[], None]:
+    """Subscribe to condition platform events."""
+    condition_platform_event_subscriptions = hass.data[CONDITION_PLATFORM_SUBSCRIPTIONS]
+
+    def remove_subscription() -> None:
+        condition_platform_event_subscriptions.remove(on_event)
+
+    condition_platform_event_subscriptions.append(on_event)
+    return remove_subscription
+
+
+async def _register_condition_platform(
+    hass: HomeAssistant, integration_domain: str, platform: ConditionProtocol
+) -> None:
+    """Register a condition platform."""
+
+    new_conditions: set[str] = set()
+
+    if hasattr(platform, "async_get_conditions"):
+        for condition_key in await platform.async_get_conditions(hass):
+            hass.data[CONDITIONS][condition_key] = integration_domain
+            new_conditions.add(condition_key)
+    else:
+        _LOGGER.debug(
+            "Integration %s does not provide condition support, skipping",
+            integration_domain,
+        )
+        return
+
+    tasks: list[asyncio.Task[None]] = [
+        create_eager_task(listener(new_conditions))
+        for listener in hass.data[CONDITION_PLATFORM_SUBSCRIPTIONS]
+    ]
+    await asyncio.gather(*tasks)
```

Comment:
If these don't ever yield to the loop its faster to await in a loop

Reply:
In the PR, listeners are added by WS command "condition_platforms/subscribe", which calls `condition.async_get_all_descriptions`.

`condition.async_get_all_descriptions` will yield in case of a cache miss in order to do file I/O, otherwise not.
Once startup is complete and platforms with conditions are all loaded, this will never yield.

However, why doesn't `asyncio.gather` eagerly run tasks? Is there no such option?

Reply:
Also since `async_get_all_descriptions` doesn't have a lock (thats fine as more efficient most of the time), you might have a thundering heard trying to build the cache if you end up in the executor because the cache is stale.  I haven't fully analyzed the flow though.